rust-runtime/aws-smithy-http-server/src/exception.rs

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Exception types.
+//!
+//! This module contains "exception types". There is only one such type, [`SmithyFrameworkException`].
+//!
+//! As opposed to rejection types (see [`crate::rejection`]), which are an internal detail about
+//! the framework, exceptions are surfaced to clients in HTTP responses: indeed, they implement
+//! [`axum_core::response::IntoResponse`]. Rejections can be "grouped" and converted into an
+//! exception type: for example, all request rejections due to serialization issues can be
+//! conflated under the [`SmithyFrameworkExceptionType::Serialization`] exception type enum
+//! variant.
+//!
+//! The HTTP response representation of an exception can be protocol-specific: for example,
+//! exceptions in the RestJson1 protocol set the `X-Amzn-Errortype` header.
+//!
+//! Exception types' behavior doesn't have anything to do with how _programming language exceptions_
+//! behave: there are no exceptions in Rust, `[SmithyFrameworkException`] is just a regular Rust
+//! enum. Their name comes from their representation in all AWS protocols: for example,
+//! [`SmithyFrameworkExceptionType::Serialization`] renders an `X-Amzn-Errortype` header with the
+//! value `SerializationException` when using the RestJson1 protocol.
+//!
+//! The way generated code works is it always works with [`crate::rejection`] types when
+//! deserializing requests and serializing response. Just before a response needs to be sent, the
+//! generated code looks up and converts into the corresponding exception type, and then it uses
+//! the exception's [`axum_core::response::IntoResponse`] implementation to render and send a
+//! response.
+
+use crate::protocols::Protocol;
+
+#[derive(Debug)]
+pub enum SmithyFrameworkExceptionType {
+    // UnknownOperation,
+    Serialization(crate::Error),
+    // UnsupportedMediaType,
+    // NotAcceptable,
```

Comment:
What about these commented values?

Reply:
They will get uncommented when I get to work on the protocol tests that exercise them; didn't want to make this PR bigger. I've copied them from the TypeScript sSDK https://github.com/awslabs/smithy-typescript/blob/2f013f6aad91851e0d4309b3ffbf90f60d3f0dd7/smithy-typescript-ssdk-libs/server-common/src/errors.ts#L18-L18 (except for the internal one, which I don't think we will need in Rust).

Reply:
IMO since they are not exercised and commented, they should just be removed until we need them..